### DISCUSSION TITLE
XmlnsDefinition don't work if add xct csproj to my project

### DISCUSSION THREAD

**Discussion by auriou** [xamarin/XamarinCommunityToolkit#1321] [Q&A] [answered]
If I add the csproj From XamarinCommunityToolkit to my project (the XCT repo is in a git submodule)

 I cannot use the references : 
xmlns:xct="http://xamarin.com/schemas/2020/toolkit"

I must use :
xmlns:xct="clr-namespace:Xamarin.CommunityToolkit.UI.Views;assembly=Xamarin.CommunityToolkit"

Otherwise I have the error 

XFC0000	Cannot resolve type "xct:........

**Comment by pictos** [comment]
@auriou are you using VSMac?

**Reply by auriou** [reply]
@pictos no i'm windows 10 pro 20H2

**Comment by GalaxiaGuy** [comment]
I recently encountered a similar problem that may be the same thing.

I tried adding a custom xmlns for the controls, effects, markup extension etc. in my app and also got the "XFC0000 Cannot resolve type..." error (although intellisense seemed to work fine if I typed my prefix in XAML).

Does a custom xmlns just not work when trying to consume it from the assembly in which it is declared? Although I guess in the case of referencing the source of XCT it would still be in a different assembly.

@pictos I was using VSMac.

**Reply by auriou** [reply]
@GalaxiaGuy It's exactly the same as I have, maybe it's a bug on xamarin 5?

Whether it's in the assembly, or a referenced it's the same

**Comment by auriou** [comment]
In the msdn doc, they say to add a call to the library initializer.

https://docs.microsoft.com/en-us/xamarin/xamarin-forms/xaml/custom-namespace-schemas#consuming-a-custom-namespace-schema

```c#
using Xamarin.Forms;
using MyCompany.Controls;

namespace CustomNamespaceSchemaDemo
{
    public partial class MainPage : ContentPage
    {
        public MainPage()
        {
            Controls.Init();
            InitializeComponent();
        }
    }
}
```
I have found several references to this problem on the internet, and the solution described is to call the initializers of the controls in the app.xaml.cs

sample :
https://supportcenter.devexpress.com/ticket/details/t992273/xamarin-forms-xfc0000-cannot-resolve-type-dxn-tabviewitem

https://github.com/roubachof/Sharpnado.Shadows/issues/24

does XCT have an initializer method ?

**Reply by pictos** [reply]
@auriou the XCT has this initializer, but we did it in a way that our users don't need to do anything, it's automatically applied. Well... Since this is just when you try to grab the assembly into your project it's fine to us. Anyway, that shouldn't happen or be a problem since we do it on our repo in the sample project, and everything works as expected.
Even with those errors you should be able to build and deploy your project

**Reply by auriou** [reply]
@pictos can you give me the name of the class and the method to initialize XCT, I did not find this call in XCT.Sample

**Reply by pictos** [reply]
you can find more info [here](https://github.com/xamarin/XamarinCommunityToolkit/pull/357)

**Reply by auriou** [reply]
@pictos  I read what you told me, but there is no Init method of setting up (Init). The problem is just if you create a .netstandard2 project and reference the XCT project and not the nuget package.

**Comment by auriou** [comment, marked as answer]
Here's how to reproduce the error

![xmlns_xct](https://user-images.githubusercontent.com/1873587/119136998-9d9e4280-ba40-11eb-8881-b84468a90b6f.gif)

for this to work, you have to add a static class which serves just as an initializer, and apply it just after the initialization of your application's initializer

```c#
// into Xamarin.CommunityToolkit project
namespace Xamarin.CommunityToolkit
{
	public static class Initializer
	{
		public static void Init()
		{
		}
	}
}

// into your project 
namespace MyApp
{
    public partial class App : Application
    {
        public App()
        {
            InitializeComponent();
            Xamarin.CommunityToolkit.Initializer.Init();
        }
    }
}
